frame/nomination-pools/src/lib.rs

```diff
@@ -0,0 +1,1525 @@
+/// `no_era	 pool. This is guaranteed to at least be equal to the staking `UnbondingDuration`. For
+/// improved UX [`Config::PostUnbondingPoolsWindow`] should be configured to a non-zero value.
+pub struct TotalUnbondingPools<T: Config>(PhantomData<T>);
+impl<T: Config> Get<u32> for TotalUnbondingPools<T> {
+	fn get() -> u32 {
+		// TODO: This may be too dangerous in the scenario bonding_duration gets decreased because
+		// we would no longer be able to decode `SubPoolsWithEra`, which uses `TotalUnbondingPools`
+		// as the bound
+		T::StakingInterface::bonding_duration() + T::PostUnbondingPoolsWindow::get()
+	}
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_system::{ensure_signed, pallet_prelude::*};
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(crate) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// Weight information for extrinsics in this pallet.
+		type WeightInfo: weights::WeightInfo;
+
+		/// The nominating balance.
+		type Currency: Currency<Self::AccountId>;
+
+		// Infallible method for converting `Currency::Balance` to `U256`.
+		type BalanceToU256: Convert<BalanceOf<Self>, U256>;
+
+		// Infallible method for converting `U256` to `Currency::Balance`.
+		type U256ToBalance: Convert<U256, BalanceOf<Self>>;
+
+		/// The interface for nominating.
+		type StakingInterface: StakingInterface<
+			Balance = BalanceOf<Self>,
+			AccountId = Self::AccountId,
+			LookupSource = <Self::Lookup as StaticLookup>::Source,
+		>;
+
+		/// The amount of eras a `SubPools::with_era` pool can exist before it gets merged into the
+		/// `SubPools::no_era` pool. In other words, this is the amount of eras a delegator will be
+		/// able to withdraw from an unbonding pool which is guaranteed to have the correct ratio of
+		/// points to balance; once the `with_era` pool is merged into the `no_era` pool, the ratio
+		/// can become skewed due to some slashed ratio getting merged in at some point.
+		type PostUnbondingPoolsWindow: Get<u32>;
+	}
+
+	/// Minimum amount to bond to join a pool.
+	#[pallet::storage]
+	pub type MinJoinBond<T: Config> = StorageValue<_, BalanceOf<T>, ValueQuery>;
+
+	/// Minimum bond required to create a pool.
+	#[pallet::storage]
+	pub type MinCreateBond<T: Config> = StorageValue<_, BalanceOf<T>, ValueQuery>;
+
+	/// Maximum number of nomination pools that can exist. If `None`, then an unbounded number of
+	/// pools can exist.
+	#[pallet::storage]
+	pub type MaxPools<T: Config> = StorageValue<_, u32, OptionQuery>;
+
+	/// Active delegators.
+	#[pallet::storage]
+	pub type Delegators<T: Config> = CountedStorageMap<_, Twox64Concat, T::AccountId, Delegator<T>>;
+
+	/// To get or insert a pool see [`BondedPool::get`] and [`BondedPool::put`]
+	#[pallet::storage]
+	pub type BondedPools<T: Config> =
```

Comment:
What you did with subpools makes more sense, and is consistent with my other works as well: call the storage item BondedPoolsStorage, and the struct it stores BondedPool(s)

Reply:
the issue is I already have `struct BondedPoolStorage` which is what gets stored. But they api used is `struct BondedPool`, which is the same as `BondedPoolStorage`, but also  includes the pool account id so we can have cleaner abstractions with the struct methods. Ideally we could just have one struct and use `#[codec(skip)]` on the account id, but since account id can't have a default we can't use `#[codec(skip)]`

Reply:
Happy though to take suggestions on different naming scheme/design. The `BondedPool` and `BondedPoolStorage` structs are annoyingly redundant